odoo/tests/common.py

```diff
@@ -1401,14 +1423,21 @@ def setUpClass(cls):
         ICP = cls.env['ir.config_parameter']
         ICP.set_param('web.base.url', cls.base_url())
         ICP.flush()
+        # v8 api with correct xmlrpc exception handling.
+        cls.xmlrpc_url = f'http://{HOST}:{odoo.tools.config["http_port"]:d}/xmlrpc/2/'
+        cls._logger = logging.getLogger('%s.%s' % (cls.__module__, cls.__name__))
 
     def setUp(self):
         super().setUp()
         if self.registry_test_mode:
             self.registry.enter_test_mode(self.cr)
             self.addCleanup(self.registry.leave_test_mode)
+
+        self.xmlrpc_common = xmlrpclib.ServerProxy(self.xmlrpc_url + 'common', transport=Transport(self.cr))
+        self.xmlrpc_db = xmlrpclib.ServerProxy(self.xmlrpc_url + 'db', transport=Transport(self.cr))
+        self.xmlrpc_object = xmlrpclib.ServerProxy(self.xmlrpc_url + 'object', transport=Transport(self.cr))
         # setup an url opener helper
-        self.opener = requests.Session()
+        self.opener = Opener(self.cr)
```

Comment:
`self.cr` is assigned by `TransactionCase.setUpClass`, and remains the same for all test methods.
Therefore all that code could be in `setUpClass`, IMO.

Reply:
These attributes were always on `self` so it felt more sensible to not touch anything and keep them there, and that way they're after the test mode bits, which could eventually be relevant (though that seems unlikely).

Reply:
Sure, these attributes were always on `self`, but that code was done *before* the refactor of test classes, which now uses `setUpClass` in a more idiomatic way.  It should already have moved to `setUpClass`, IMO.  I can't remember whether there was a good reason for not doing it (besides laziness).

Reply:
> IMO. I can't remember whether there was a good reason for not doing it (besides laziness).

As added in the edit (which you may have not seen) `opener` has to be reset between tests as it gets updated in-place by authentication, not resetting it would lead to shared session ids between tests, and incoherences between the Session and the browser.

For xmlrpc it's less relevant as they're mostly stateless, but it seems logical to keep the thing in sync with the opener, and mostly I saw that `Transport` has ` `connection` so figured ensuring everything was reset between tests would be more prudent.